In [47]:
import pandas as pd
import pypyodbc as podbc
import numpy as np
from sklearn.feature_extraction import DictVectorizer
from sklearn import linear_model
from scipy.sparse import csr_matrix
from collections import OrderedDict

In [2]:
#pandas params
pd.set_option('display.max_rows', None)
pd.set_option('display.min_rows', 200)
pd.set_option('display.max_columns', 200)
pd.set_option('display.max_colwidth', 200)

In [434]:
#Connect to RDS db
conn = podbc.connect(
    #Trusted_Connection='Yes',
    Driver='{SQL Server}',
    Server='nbahistorical.cmgzaupuq9dz.us-east-2.rds.amazonaws.com',
    UID='admin',
    PWD='Gs239350',
    Database='pbp',
    MultipleActiveResultSets = 'True'
)

In [5]:
df = pd.read_sql_query(sql = """
SELECT
	* FROM (
SELECT 
	EVENTNUM AS E, 
	HOMEDESCRIPTION,
	VISITORDESCRIPTION,
	EVENTMSGTYPE,
	GAME_ID,
	STRING_AGG(PERSON_ID, ', ') 
		WITHIN GROUP (ORDER BY TEAM_ID,PERSON_ID)
	AS STINT,
	ROW_NUMBER () 
		OVER (PARTITION BY GAME_ID, EVENTNUM ORDER BY GAME_ID) AS R 
FROM
		(
		SELECT  
			*,
			(CASE
			--Method for time during regulation except at beginning of quarter, with >= 10 minutes on the clock
			WHEN PERIOD <5 AND PCTIMESTRING NOT LIKE '12:0%' AND PCTIMESTRING LIKE '__:%'
			THEN
			(((CAST(period AS int) - 1) * 7200)
					+ 
					(((CAST(SUBSTRING(PCTIMESTRING,1,2) AS int)
						-11) 
							*(-1))
								*600)
					+
					((CAST(SUBSTRING(PCTIMESTRING,4,2) AS int)
						-60)
							*(-10)))
			--Method for time during regulation except at beginning of quarter, with < 10 minutes on the clock
			WHEN PERIOD <5 AND PCTIMESTRING NOT LIKE '12:0%' AND PCTIMESTRING NOT LIKE '__:%'
			THEN
			(((CAST(period AS int) - 1) * 7200)
					+ 
					(((CAST(SUBSTRING(PCTIMESTRING,1,1) AS int)
						-11) 
							*(-1))
								*600)
					+
					((CAST(SUBSTRING(PCTIMESTRING,4,2) AS int)
						-60)
							*(-10)))
		--Method for time during regulation at beginning of quarter only
			WHEN PERIOD <5 AND PCTIMESTRING LIKE '12:0%'
			THEN
			(
			(
				((CAST(period AS int)) - 1)
				* 7200)
					+ (
					((((CAST(SUBSTRING(PCTIMESTRING,1,2) AS int))
						-12)
							*(-1))
								*60)
				)
			)	
		--Method for time during overtime 
			WHEN PERIOD >4
			THEN
			(28800 +
			(((CAST(period AS int)) - 5) * 3000)
					+ 
					((((CAST(SUBSTRING(PCTIMESTRING,1,1) AS int))
						-4)
							*(-1))
								*600)
					+
					(((CAST(SUBSTRING(PCTIMESTRING,4,2) AS int))
						-60)
							*(-10))
				)
			ELSE 'ERROR'
			END)
			AS TIME
		FROM 	
			all_pbp
		)
		play_by_play
			LEFT JOIN
				(
				SELECT
					GAME_ID AS GAME,
					CAST(TEAM_ID AS int) AS TEAM_ID,
					TEAM_CITY,
					TEAM_NAME,
					PERSON_ID,
					PLAYER_NAME,
					CAST(TIME_IN AS int) AS TIME_IN,
					CAST(TIME_OUT AS int) AS TIME_OUT,
					PLAYER_POINTS,
					POINT_MARGIN,
					USAGE
				FROM
					rotations
				) b
					ON play_by_play.GAME_ID = b.GAME
					AND TIME < TIME_OUT
					AND TIME >= TIME_IN
	GROUP BY 
		GAME_ID,
		PERIOD,
		EVENTNUM,
		HOMEDESCRIPTION,
		VISITORDESCRIPTION,
		EVENTMSGTYPE
) q
WHERE 
	GAME_ID LIKE '215%'
	AND R = 1
;
""", con = conn)

In [6]:
df.to_csv(f"C:\\Users\gsteele\Other\\2015_16_pbp.csv")

In [4]:
df = pd.read_csv(f"C:\\Users\gsteele\Other\\2015_16_pbp.csv")

In [5]:
pbp = pd.read_sql_query(sql = """
SELECT 
   GAME_ID, EVENTNUM, EVENTMSGTYPE, PERIOD, PCTIMESTRING, HOMEDESCRIPTION, VISITORDESCRIPTION, SCORE, SCOREMARGIN, 
   PLAYER1_ID, PLAYER1_TEAM_ID
FROM 
    all_pbp
WHERE
    GAME_ID LIKE '215%';
""", con = conn)

In [6]:
pbp.columns = map(str.upper, pbp.columns)

#Converting EVENTNUM to int, then creating true event number to cover gaps
pbp['EVENTNUM'] = pbp['EVENTNUM'].astype(int)
pbp['EVENT'] = pbp.groupby(['GAME_ID'])['EVENTNUM'].rank()

#Manipulating SCOREMARGIN
pbp['SCOREMARGIN'] = pbp['SCOREMARGIN'].replace('',np.nan)
pbp['SCOREMARGIN'] = pbp.groupby(['GAME_ID'])['SCOREMARGIN'].fillna(method = 'ffill')
pbp['SCOREMARGIN'] = np.where(pbp['SCOREMARGIN'] == 'TIE', 0, pbp['SCOREMARGIN'])
pbp['SCOREMARGIN'] = pbp['SCOREMARGIN'].fillna(0)

#Manipulating SCORE
pbp['SCORE'] = pbp['SCORE'].replace('',np.nan)
pbp['SCORE'] = pbp.groupby(['GAME_ID'])['SCORE'].fillna(method = 'ffill')
pbp['SCORE'] = pbp['SCORE'].fillna('0 - 0')

#Create HOME_SCORE and AWAY_SCORE
pbp.insert(loc = 11, column = 'HOME_SCORE', value = [i[1] for i in pbp['SCORE'].str.split(' - ')])
pbp.insert(loc = 12, column = 'AWAY_SCORE', value = [i[0] for i in pbp['SCORE'].str.split(' - ')])

#Create NEXT_EVENTNUM, join associated data for next event to current event  
pbp.insert(loc = 13, column = 'NEXT_EVENTNUM', value = pbp['EVENT'].astype(int) + 1)
event = pbp[['GAME_ID','EVENT','HOMEDESCRIPTION','VISITORDESCRIPTION']]
#event['EVENTNUM'] = event['EVENTNUM'].astype(int)
event = event.rename(columns = {"EVENT":"NEXT_EVENTNUM",#"EVENT":"NEXT_EVENT",
                                "HOMEDESCRIPTION":"NEXT_HOMEDESCRIPTION","VISITORDESCRIPTION":"NEXT_VISITORDESCRIPTION"})
pbp = pbp.merge(right = event, how = 'left', on = ['GAME_ID','NEXT_EVENTNUM'])

#Repeat the process for the preceding event
pbp.insert(loc = 14, column = 'LAST_EVENTNUM', value = pbp['EVENT'].astype(int) - 1)
event = event.rename(columns = {"NEXT_EVENTNUM":"LAST_EVENTNUM",#"NEXT_EVENT":"LAST_EVENT",
                               "NEXT_HOMEDESCRIPTION":"LAST_HOMEDESCRIPTION","NEXT_VISITORDESCRIPTION":"LAST_VISITORDESCRIPTION"})
#pbp = pbp.merge(right = event, how = 'left', on = [['GAME_ID','LAST_EVENTNUM']])
pbp = pbp.merge(right = event, how = 'left', left_on = ['GAME_ID','LAST_EVENTNUM'], right_on = ['GAME_ID','LAST_EVENTNUM'])

C:\Users\gsteele\Miniconda3\lib\site-packages\pandas\core\reshape\merge.py:1123: UserWarning: You are merging on int and float columns where the float values are not equal to their int representation
  warnings.warn(


In [7]:
#Transform time to be consistent with rotations data
pbp.insert(loc = 16, column = 'MINUTES', value = [i[0] for i in pbp['PCTIMESTRING'].str.split(':')])
pbp.insert(loc = 16, column = 'SECONDS', value = [i[1] for i in pbp['PCTIMESTRING'].str.split(':')])

#Transform time suring regulation except at beginning of period
pbp['TIME'] = np.where(((pbp['PERIOD'].astype(int) < 5) & (pbp['PCTIMESTRING'] != '12:00')), 
                      (((pbp['PERIOD'].astype(int) -1) *7200) 
                      + ((pbp['MINUTES'].astype(int) -11) * (-600))
                      + ((pbp['SECONDS'].astype(int) -60) * (-10))),
                          pbp['PCTIMESTRING'])
#Transform time during regulation at beginning of period
pbp['TIME'] = np.where(((pbp['PERIOD'].astype(int) < 5) & (pbp['PCTIMESTRING'] == '12:00')), 
                      (((pbp['PERIOD'].astype(int) -1) *7200) 
                      + ((pbp['MINUTES'].astype(int) -12) * (-600))),
                          pbp['TIME'])
#Transform time during overtime
pbp['TIME'] = np.where(pbp['PERIOD'].astype(int) > 4, 
                      (28800 + 
                      ((pbp['PERIOD'].astype(int) -5) *3000) 
                      + ((pbp['MINUTES'].astype(int) -4) * (-600))
                      + ((pbp['SECONDS'].astype(int) -60) * (-10))),
                          pbp['TIME'])

The following anomaly is the result of two records having the same GAME_ID and EVENTNUM, resulting in a rank() output that is a float (e.g., 380.5). Need to research ways to implement a dense rank and achieve the same effect. 

In [8]:
len(pbp[(pbp['NEXT_HOMEDESCRIPTION'].isnull() == True)
    & (pbp['TIME'] != 28800) & (pbp['PCTIMESTRING'] != '0:00')])
#pbp.sort_values(by = ['GAME_ID','EVENT']).head()

296

In [14]:
df['e'] = df['e'].astype(int)
#Needed when pulling from csv
df['game_id'] = df['game_id'].astype(str)
pbp_df = df.merge(right = pbp, how = 'left', left_on = ['game_id','e'], right_on = ['GAME_ID','EVENTNUM'])

In [768]:
#len(pbp_df[pbp_df['stint'].isna() == True])
pbp_df['DESCRIPTION'] = (pbp_df['HOMEDESCRIPTION'] + pbp_df['VISITORDESCRIPTION'])
pbp_df['NEXT_DESCRIPTION'] = (pbp_df['NEXT_HOMEDESCRIPTION'] + pbp_df['NEXT_VISITORDESCRIPTION'])
pbp_df['LAST_DESCRIPTION'] = (pbp_df['LAST_HOMEDESCRIPTION'] + pbp_df['LAST_VISITORDESCRIPTION'])

#Away ball
pbp_df['HA'] = np.where(((pbp_df['HOMEDESCRIPTION'] == '') 
                        | (pbp_df['HOMEDESCRIPTION'].isna())) # added
                        & (pbp_df['EVENTMSGTYPE'] != '2') 
                        & (pbp_df['EVENTMSGTYPE'] != '5'), 'A', 0)
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '5') & (pbp_df['HOMEDESCRIPTION'].str.contains('STEAL')), 
                       'A', pbp_df['HA'])
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') & (pbp_df['HOMEDESCRIPTION'].str.contains('BLOCK')), 
                       'A', pbp_df['HA'])  
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') & (pbp_df['HOMEDESCRIPTION'] == ''), 
                       'A', pbp_df['HA']) 
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '5') & (pbp_df['HOMEDESCRIPTION'] == ''), 
                       'A', pbp_df['HA']) 
pbp_df['HA'] = np.where(pbp_df['HOMEDESCRIPTION'].str.contains('S.FOUL'), 'A', pbp_df['HA'])

#Home ball
pbp_df['HA'] = np.where((pbp_df['VISITORDESCRIPTION'] == '') 
                        & (pbp_df['EVENTMSGTYPE'] != '2') 
                        & (pbp_df['EVENTMSGTYPE'] != '5'), 'H', pbp_df['HA'])
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '5') & (pbp_df['VISITORDESCRIPTION'].str.contains('STEAL')), 
                       'H', pbp_df['HA'])
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') & (pbp_df['VISITORDESCRIPTION'].str.contains('BLOCK')), 
                       'H', pbp_df['HA'])  
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') & (pbp_df['VISITORDESCRIPTION'] == ''), 
                       'H', pbp_df['HA'])  
pbp_df['HA'] = np.where((pbp_df['EVENTMSGTYPE'] == '5') & (pbp_df['VISITORDESCRIPTION'] == ''), 
                       'H', pbp_df['HA'])  
pbp_df['HA'] = np.where(pbp_df['VISITORDESCRIPTION'].str.contains('S.FOUL'), 'H', pbp_df['HA'])

###Counting Possessions
#Turnovers and violations
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '5') | (pbp_df['EVENTMSGTYPE'] == '7'), 1, 0)
#Made basket without a shooting foul
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '1') 
    & (pbp_df['HOMEDESCRIPTION'] == '')
   & (pbp_df['NEXT_HOMEDESCRIPTION'].str.contains('S.FOUL') == False), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '1') 
    & (pbp_df['VISITORDESCRIPTION'] == '')
   & (pbp_df['NEXT_VISITORDESCRIPTION'].str.contains('S.FOUL') == False), 1, pbp_df['POSS'])
#Made final free throw
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '3') 
                          & (pbp_df['DESCRIPTION'].str.contains('1 of 1'))
                          & (pbp_df['DESCRIPTION'].str.contains('MISS') == False), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '3') 
                          & (pbp_df['DESCRIPTION'].str.contains('2 of 2'))
                          & (pbp_df['DESCRIPTION'].str.contains('MISS') == False), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '3') 
                          & (pbp_df['DESCRIPTION'].str.contains('3 of 3'))
                          & (pbp_df['DESCRIPTION'].str.contains('MISS') == False), 1, pbp_df['POSS'])
#Defensive rebound
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') 
                          & (pbp_df['HOMEDESCRIPTION'] == '')
                          & (pbp_df['NEXT_HOMEDESCRIPTION'].str.contains('REBOUND')), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') 
                          & (pbp_df['VISITORDESCRIPTION'] == '')
                          & (pbp_df['NEXT_VISITORDESCRIPTION'].str.contains('REBOUND')), 1, pbp_df['POSS'])
#Team defensive rebound
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') 
                          & (pbp_df['HOMEDESCRIPTION'] == '')
                          & (pbp_df['NEXT_HOMEDESCRIPTION'].str.contains('Rebound')), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') 
                          & (pbp_df['VISITORDESCRIPTION'] == '')
                          & (pbp_df['NEXT_VISITORDESCRIPTION'].str.contains('Rebound')), 1, pbp_df['POSS'])
#Block defensive rebounds
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') 
                          & (pbp_df['HOMEDESCRIPTION'].str.contains('BLOCK'))
                          & (pbp_df['NEXT_HOMEDESCRIPTION'].str.contains('REBOUND')), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '2') 
                          & (pbp_df['VISITORDESCRIPTION'].str.contains('BLOCK'))
                          & (pbp_df['NEXT_VISITORDESCRIPTION'].str.contains('REBOUND')), 1, pbp_df['POSS'])
#Free throw rebounds
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '3') 
                          & (pbp_df['DESCRIPTION'].str.contains('MISS'))
                          & (pbp_df['NEXT_HOMEDESCRIPTION'].str.contains('REBOUND'))
                          & (pbp_df['HOMEDESCRIPTION'] == ''), 1, pbp_df['POSS'])
pbp_df['POSS'] = np.where((pbp_df['EVENTMSGTYPE'] == '3') 
                          & (pbp_df['DESCRIPTION'].str.contains('MISS'))
                          & (pbp_df['NEXT_VISITORDESCRIPTION'].str.contains('REBOUND'))
                          & (pbp_df['VISITORDESCRIPTION'] == ''), 1, pbp_df['POSS'])
pbp_df['STINT_POSS'] = pbp_df.groupby(['GAME_ID','PERIOD','stint','HA'])['POSS'].transform('sum')

###Calculating Points
pbp_df['PTS'] = np.where((pbp_df['EVENTMSGTYPE'] == '3') & (pbp_df['DESCRIPTION'].str.contains('MISS') == False), 1, 0)
pbp_df['PTS'] = np.where((pbp_df['EVENTMSGTYPE'] == '1') 
                         & (pbp_df['DESCRIPTION'].str.contains('3PT Jump')), 3, pbp_df['PTS'])
pbp_df['PTS'] = np.where((pbp_df['EVENTMSGTYPE'] == '1') 
                         & (pbp_df['DESCRIPTION'].str.contains('3PT Jump') == False), 2, pbp_df['PTS'])
pbp_df['STINT_PTS'] = pbp_df.groupby(['GAME_ID','PERIOD','stint','HA'])['PTS'].transform('sum')

#Team ids for home and away
pbp_df['HOME_TEAM'] = np.where((pbp_df['HOMEDESCRIPTION'] != '')
                               & (pbp_df['VISITORDESCRIPTION'] == '')
                               & (pbp_df['HOMEDESCRIPTION'] != 0)
                               & (pbp_df['HOMEDESCRIPTION'].isna() == False)
                               & (pbp_df['HOMEDESCRIPTION'].isnull() == False), pbp_df['PLAYER1_TEAM_ID'], 0)
pbp_df['AWAY_TEAM'] = np.where((pbp_df['VISITORDESCRIPTION'] != '')
                               & (pbp_df['HOMEDESCRIPTION'] == '')
                               & (pbp_df['VISITORDESCRIPTION'] != 0)
                               & (pbp_df['VISITORDESCRIPTION'].isna() == False)
                               & (pbp_df['VISITORDESCRIPTION'].isnull() == False), pbp_df['PLAYER1_TEAM_ID'], 0)

In [773]:
home_team = pbp_df[['game_id','HOME_TEAM']].drop_duplicates()
home_team = home_team[home_team['HOME_TEAM'] != 0]
home_team = home_team[home_team['HOME_TEAM'] != '']
away_team = pbp_df[['game_id','AWAY_TEAM']].drop_duplicates()
away_team = away_team[away_team['AWAY_TEAM'] != 0]
away_team = away_team[away_team['AWAY_TEAM'] != '']
team_dimension_table = home_team.merge(right = away_team, how = 'inner', on = ['game_id'])

In [774]:
pbp_df = pbp_df[pbp_df['stint'].isna() == False]

In [776]:
pbp_df[(pbp_df['STINT_POSS'] == 0) & (pbp_df['STINT_PTS'] > 0)].head()

Unnamed: 0    e                 homedescription  \
2995         2995  175    Jackson 6' Jump Shot (4 PTS)   
2996         2996  176                             NaN   
2998         2998  179  MISS Jackson Free Throw 1 of 1   
16448       16447  466                             NaN   
16449       16448  468                             NaN   

                       visitordescription  eventmsgtype   game_id  \
2995                                  NaN             1  21500007   
2996     Neto S.FOUL (P1.T3) (J.VanDuyne)             6  21500007   
2998                                  NaN             3  21500007   
16448  Williams Free Throw 1 of 2 (6 PTS)             3  21500034   
16449     MISS Williams Free Throw 2 of 2             3  21500034   

                                                                                 stint  \
2995   202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   
2996   202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   
2998   202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   
16448    1626170, 201577, 202498, 202682, 204038, 201196, 201971, 202390, 203078, 2400   
16449    1626170, 201577, 202498, 202682, 204038, 201196, 201971, 202390, 203078, 2400   

       r   GAME_ID  EVENTNUM EVENTMSGTYPE PERIOD PCTIMESTRING  \
2995   1  21500007       175            1      2         3:49   
2996   1  21500007       176            6      2         3:49   
2998   1  21500007       179            3      2         3:49   
16448  1  21500034       466            3      3         0:50   
16449  1  21500034       468            3      3         0:50   

                      HOMEDESCRIPTION                  VISITORDESCRIPTION  \
2995     Jackson 6' Jump Shot (4 PTS)                                       
2996                                     Neto S.FOUL (P1.T3) (J.VanDuyne)   
2998   MISS Jackson Free Throw 1 of 1                                       
16448                                  Williams Free Throw 1 of 2 (6 PTS)   
16449                                     MISS Williams Free Throw 2 of 2   

         SCORE SCOREMARGIN PLAYER1_ID PLAYER1_TEAM_ID HOME_SCORE AWAY_SCORE  \
2995   36 - 31          -5     202704    1610612765.0         31         36   
2996   36 - 31          -5     203526    1610612762.0         31         36   
2998   36 - 31          -5     202704    1610612765.0         31         36   
16448  86 - 85          -1     202682    1610612752.0         85         86   
16449  86 - 85          -1     202682    1610612752.0         85         86   

       NEXT_EVENTNUM  LAST_EVENTNUM  EVENT SECONDS MINUTES  \
2995             163            161  162.0      49       3   
2996             164            162  163.0      49       3   
2998             166            164  165.0      49       3   
16448            378            376  377.0      50       0   
16449            379            377  378.0      50       0   

      NEXT_HOMEDESCRIPTION           NEXT_VISITORDESCRIPTION  \
2995                        Neto S.FOUL (P1.T3) (J.VanDuyne)   
2996                                     SUB: Burks FOR Neto   
2998                                            Jazz Rebound   
16448                        MISS Williams Free Throw 2 of 2   
16449                                         Knicks Rebound   

                  LAST_HOMEDESCRIPTION             LAST_VISITORDESCRIPTION  \
2995                                             Hood 5' Jump Shot (6 PTS)   
2996      Jackson 6' Jump Shot (4 PTS)                                       
2998                                                   SUB: Burks FOR Neto   
16448  Gooden S.FOUL (P2.PN) (K.Scott)                                       
16449                                   Williams Free Throw 1 of 2 (6 PTS)   

        TIME                         DESCRIPTION  \
2995   12110        Jackson 6' Jump Shot (4 PTS)   
2996   12110    Neto S.FOUL (P1.T3) (J.Van

In [778]:
pbp_df[(pbp_df['game_id'] == '21500007')
    & (pbp_df['stint'] == '202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484')
    & (pbp_df['PERIOD'] == '2')]

Unnamed: 0    e                 homedescription  \
2995        2995  175    Jackson 6' Jump Shot (4 PTS)   
2996        2996  176                             NaN   
2997        2997  178                             NaN   
2998        2998  179  MISS Jackson Free Throw 1 of 1   

                    visitordescription  eventmsgtype   game_id  \
2995                               NaN             1  21500007   
2996  Neto S.FOUL (P1.T3) (J.VanDuyne)             6  21500007   
2997               SUB: Burks FOR Neto             8  21500007   
2998                               NaN             3  21500007   

                                                                                stint  \
2995  202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   
2996  202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   
2997  202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   
2998  202324, 202330, 202407, 202692, 203918, 1626169, 202694, 202704, 203083, 203484   

      r   GAME_ID  EVENTNUM EVENTMSGTYPE PERIOD PCTIMESTRING  \
2995  1  21500007       175            1      2         3:49   
2996  1  21500007       176            6      2         3:49   
2997  1  21500007       178            8      2         3:49   
2998  1  21500007       179            3      2         3:49   

                     HOMEDESCRIPTION                VISITORDESCRIPTION  \
2995    Jackson 6' Jump Shot (4 PTS)                                     
2996                                  Neto S.FOUL (P1.T3) (J.VanDuyne)   
2997                                               SUB: Burks FOR Neto   
2998  MISS Jackson Free Throw 1 of 1                                     

        SCORE SCOREMARGIN PLAYER1_ID PLAYER1_TEAM_ID HOME_SCORE AWAY_SCORE  \
2995  36 - 31          -5     202704    1610612765.0         31         36   
2996  36 - 31          -5     203526    1610612762.0         31         36   
2997  36 - 31          -5     203526    1610612762.0         31         36   
2998  36 - 31          -5     202704    1610612765.0         31         36   

      NEXT_EVENTNUM  LAST_EVENTNUM  EVENT SECONDS MINUTES  \
2995            163            161  162.0      49       3   
2996            164            162  163.0      49       3   
2997            165            163  164.0      49       3   
2998            166            164  165.0      49       3   

                NEXT_HOMEDESCRIPTION           NEXT_VISITORDESCRIPTION  \
2995                                  Neto S.FOUL (P1.T3) (J.VanDuyne)   
2996                                               SUB: Burks FOR Neto   
2997  MISS Jackson Free Throw 1 of 1                                     
2998                                                      Jazz Rebound   

              LAST_HOMEDESCRIPTION           LAST_VISITORDESCRIPTION   TIME  \
2995                                       Hood 5' Jump Shot (6 PTS)  12110   
2996  Jackson 6' Jump Shot (4 PTS)                                    12110   
2997                                Neto S.FOUL (P1.T3) (J.VanDuyne)  12110   
2998                                             SUB: Burks FOR Neto  12110   

                           DESCRIPTION                  NEXT_DESCRIPTION  \
2995      Jackson 6' Jump Shot (4 PTS)  Neto S.FOUL (P1.T3) (J.VanDuyne)   
2996  Neto S.FOUL (P1.T3) (J.VanDuyne)               SUB: Burks FOR Neto   
2997               SUB: Burks FOR Neto    MISS Jackson Free Throw 1 of 1   
2998    MISS Jackson Free Throw 1 of 1                      Jazz Rebound   

                      LAST_DESCRIPTION HA  POSS  STINT_POSS  PTS  STINT_PTS  \
2995         Hood 5' Jump Shot (6 PTS)  H     0           0    2          2   
2996      Jackson 6' Jump Shot (4 PTS)  H     0           0    0          2   
2997  Neto S.FOUL (P1.T3) (J.VanDuyne)  A     0           0    0          0   
2998               SUB: Burks FOR Neto  H     0           0    0          2   

         H

In [779]:
def stint_deduplicater(s):  
    if len([i for i in s.split(', ')]) == 20:
        j = [i for i in s.split(', ')]
        k = j[0] + ', ' + j[2] + ', ' + j[4] + ', ' + j[6] + ', ' + j[8] + ', ' + j[10] + ', ' + j[12] + ', ' + j[14] + ', ' + j[16] + ', ' + j[18]
        return k    
    else:
        return s

In [780]:
pbp_df['stint'] = pbp_df['stint'].map(stint_deduplicater)

#testing the function
for i in pbp_df['stint']:
    if len([j for j in i.split(', ')]) == 20:
        print(i)

In [781]:
def find_error_rows(s):
    if len([i for i in s.split(', ')]) > 10:
        return 'Error'    
    else:
        return s

In [782]:
pbp_df['stint'] = pbp_df['stint'].map(find_error_rows)

#testing the function
for i in pbp_df['stint']:
    if len([j for j in i.split(', ')]) > 10:
        print(i)

In [785]:
print(pbp_df[pbp_df['stint'] == 'Error']['game_id'].value_counts())
pbp_df = pbp_df[pbp_df['stint'] != 'Error']

In [31]:
#pbp_df[pbp_df['stint'] == '101123, 1626159, 2547, 2617, 2757, 201945, 202334, 203081, 203086, 203459, 203468, 203486, 203943']
#pbp_df[(pbp_df['game_id'] == '21500404') & (pbp_df['PERIOD'] == '1')].sort_values(['e'])

In [729]:
#stints[stints['ORATING'].isna()].head(10)
#stints[(stints['stint'] == '201567, 202389, 203099, 2544, 2590, 201565, 201959, 202710, 203503, 2200')
#      & (stints['game_id'] == '21500002')]
#[(stints['game_id'] == '21500002') & (stints['PERIOD'] == '1')]

#len(stints[(stints['STINT_PTS'] > 0) & (stints['STINT_POSS'] == 0)])
#pbp_df[(pbp_df['stint'] == '201567, 202389, 203099, 2544, 2590, 201565, 201959, 202710, 203503, 2200')
#      & (pbp_df['game_id'] == '21500002')]

In [786]:
stints = pbp_df[['game_id','stint','PERIOD','HA','STINT_POSS','STINT_PTS']].drop_duplicates()
stints = stints[stints['stint'].isna() == False]
stints = stints.merge(right = team_dimension_table, how = 'left', on = ['game_id'])
stints['ORATING'] = stints['STINT_PTS'] / stints['STINT_POSS'] * 100
stints['HOME_TEAM'] = stints['HOME_TEAM'].astype(float).astype(int)
stints['AWAY_TEAM'] = stints['AWAY_TEAM'].astype(float).astype(int)

In [787]:
#If the team with the higher TEAM_ID number has the ball
stints['ordered_stint'] = np.where((stints['HOME_TEAM'] > stints['AWAY_TEAM']) 
                                   & (stints['HA'] == 'H') 
                                   & (stints['STINT_POSS'] >= 1),
                                       [s[::-1] for s in [i.split(', ') for i in stints['stint']]],
                                       stints['stint']
                                  )

stints['ordered_stint'] = np.where((stints['AWAY_TEAM'] > stints['HOME_TEAM']) 
                                   & (stints['HA'] == 'A') 
                                   & (stints['STINT_POSS'] >= 1),
                                       [s[::-1] for s in [i.split(', ') for i in stints['stint']]],
                                       stints['stint']
                                  )
stints['which_type'] = stints['ordered_stint'].apply(lambda x: str(type(x)))
stints['ordered_stint'] = np.where(stints['which_type'].str.contains('list'),
                                  stints['ordered_stint'].apply(lambda x: ', '.join(str(i) for i in x)),

                                   stints['ordered_stint'])
#Function to create the columns we need
def stint_slicer(x, n1, n2):
    return str(x.split(', ')[n1:n2]).strip("[]'")


stints['offense_1'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,0,1))
stints['offense_2'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,1,2))
stints['offense_3'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,2,3))
stints['offense_4'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,3,4))
stints['offense_5'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,4,5))
stints['defense_1'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,5,6))
stints['defense_2'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,6,7))
stints['defense_3'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,7,8))
stints['defense_4'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,8,9))
stints['defense_5'] = stints['ordered_stint'].apply(lambda x: stint_slicer(x,9,10))

stints = stints[['ordered_stint','STINT_POSS','STINT_PTS','ORATING','offense_1','offense_2','offense_3','offense_4','offense_5',
                'defense_1','defense_2','defense_3','defense_4','defense_5']]

stints.replace('', np.nan, inplace = True)
stints.dropna(subset = ['ordered_stint','STINT_POSS','STINT_PTS','offense_1','offense_2','offense_3','offense_4',
                        'offense_5','defense_1','defense_2','defense_3','defense_4','defense_5'], inplace = True)
print(len(stints))

69815


In [676]:
#Proof that reversal worked
#stints[stints['stint'] == '203484, 203083, 202704, 202694, 101141, 2594, 203145, 201952, 201143, 200794']

In [799]:
#Momentarily hard-coding records where ORating is Nan or inf
stints['ORATING'] = np.where((stints['STINT_POSS'] == 0) & (stints['STINT_PTS'] > 0),
                             (stints['STINT_PTS'] * 100),
                             stints['ORATING'])
stints = stints[stints['ORATING'].isna() == False]
len(stints)

64420

In [800]:
set_of_players = {'200794','203484'}
for col in lineups_only.columns:
    c = stints[col].tolist()
    unique_in_c = set(c)
    set_of_players = set_of_players.union(unique_in_c)
        
set_of_players = [p for p in set_of_players if p != '']
num_of_players = len(set_of_players)

In [801]:
lineups_only = stints.drop(['ordered_stint','STINT_POSS','STINT_PTS','ORATING'], axis = 1)

In [802]:
def row_processing(lineups_only, set_of_players):
    player_1 = lineups_only[0]
    player_2 = lineups_only[1]
    player_3 = lineups_only[2]
    player_4 = lineups_only[3]
    player_5 = lineups_only[4]
    player_6 = lineups_only[5]
    player_7 = lineups_only[6]
    player_8 = lineups_only[7]
    player_9 = lineups_only[8]
    player_10 = lineups_only[9]

    mat = np.zeros(num_of_players * 2)
    mat[set_of_players.index(player_1)] = 1
    mat[set_of_players.index(player_2)] = 1
    mat[set_of_players.index(player_3)] = 1
    mat[set_of_players.index(player_4)] = 1
    mat[set_of_players.index(player_5)] = 1
    mat[set_of_players.index(player_6) + num_of_players] = -1
    mat[set_of_players.index(player_7) + num_of_players] = -1
    mat[set_of_players.index(player_8) + num_of_players] = -1
    mat[set_of_players.index(player_9) + num_of_players] = -1
    mat[set_of_players.index(player_10) + num_of_players] = -1
    
    return mat

X = np.apply_along_axis(row_processing, 1, lineups_only,set_of_players)

In [791]:
X[792]

array([ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,
        0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.,  0

NEXT TO DO: 
<br>[X] Create variable to store number of players
<br>[X] Create np.zeroes with num_of_player * 2
<br>[X] Assign a 1 for offensive players at their position in the array and a -1 for defensive players at their position + num_of_players


<br>Reset stints to get of rows where STINT_POSS == 0
<br>Clean rows where STINT_PTS > 0 and STINT_POS == 0
<br>Once these are clean then:
<br>fit model

In [916]:
#clf = linear_model.RidgeCV(alphas = (np.array([0.01, 0.1, 1.0, 10, 100, 500, 750])), cv = 5)
lambdas = [.01, .05, .1]
samples = X.shape[0]
alphas = [(l * samples / 2) for l in lambdas]

clf = linear_model.RidgeCV(alphas = alphas, cv = 5, fit_intercept = True)
#clf = linear_model.RidgeCV(alphas = (np.array([0.01, 0.1, 1.0, 10, 100, 500, 750, 1000, 2000])), cv = 5)
model = clf.fit(X, stints['ORATING'], sample_weight = stints['STINT_POSS'])

In [917]:
print(clf.alpha_)

3221.0


In [805]:
roster = pd.read_sql_query(sql = """
SELECT 
   DISTINCT PERSON_ID, PLAYER_NAME
FROM 
    rotations;
""", con = conn)

In [918]:
offensive = np.array(model.coef_.T[0:num_of_players])
defensive = np.array(model.coef_.T[num_of_players:])
ids = np.array(set_of_players)
cols = np.array([ids,offensive, defensive])
coefficients = pd.DataFrame(cols).T
coefficients = coefficients.rename(columns = {0:"ID",1:"Offense",2:"Defense"})
RAPM = coefficients.merge(right = roster, how = 'left', left_on = 'ID', right_on = 'person_id')
RAPM['RAPM'] = (RAPM['Offense'].astype(float) + RAPM['Defense'].astype(float))

RAPM.sort_values(['RAPM'], ascending = False)

ID                Offense                 Defense person_id  \
377     2738      2.372479664174307      1.8268387831841373      2738   
122   201976     1.6219920288601557      2.5390814290393435    201976   
273   201601     1.4729010679632673      2.5002407791480556    201601   
74    202681     2.3520153237018886      1.4475604524267742    202681   
292   201935      2.381432349183132      0.6890278728587774    201935   
62    202695     0.7800040957383322      2.1086648904985967    202695   
69    203501     1.7417436176440444      1.1071333068976905    203501   
190   202339       1.14730713289835      1.6142401465603655    202339   
464   202083      3.265793057956895     -0.5101880447829351    202083   
385   201988     1.7271628782786854      0.8855297411620973    201988   
127     2544      2.721056671590685     -0.1134952329659457      2544   
64    203500  -0.019700861364912142      2.5675531001858474    203500   
357   203110     0.9111071074760206      1.5705511781480461    203110   
90    200755     1.1644904173373103      1.3010052607330984    200755   
417   202734     1.7474094700298066      0.6928294248205936    202734   
172   202691     1.6104118567906769      0.8291374318706981    202691   
269   201939     1.2821994391838252      1.1337457341570376    201939   
424     2555     0.5386188417237703      1.8756468131400725      2555   
448   201959     1.0147430627964436      1.2987560764411905    201959   
155  1626159    -0.2440297116852098      2.4964044822919957   1626159   
340   201950     1.2735235601943518      0.9606652202307595    201950   
123   202340     1.6674988912320545      0.5305128598774035    202340   
146   202738     2.2126697861639397    -0.02212390343271993    202738   
391   203087   -0.27047494633067265      2.4309584643335556    203087   
373   101112     1.2790083060596766      0.8060871138184804    101112   
53      1495   -0.07526492754995383      2.1487761108798433      1495   
101   202335      2.294618139706996     -0.2756932824193811    202335   
175   201196      0.979417022066955      1.0313723869996791    201196   
193     2446     0.8442338949896347      1.1636116199511233      2446   
116      708   -0.23246701723762067      2.2101411171902243       708   
270   203468     1.0934492309512376      0.8473250718975921    203468   
210   202331     1.0192109699035774      0.8877009495357056    202331   
445   201566       2.35860258718284     -0.5117364128364538    201566   
171  1626204    0.10770448453408567      1.7377227927173007   1626204   
215     2588     1.0893975286401643      0.7401963027644439      2588   
395   201147    0.25846516530688585      1.5494166843369532    201147   
236   203497   -0.41262271187811855       2.194791269653974    203497   
341   202498     0.7659046767264661       0.925751491532671    202498   
281   201949     0.7515019484111584      0.9341038926577147    201949   
289   203526     0.1969650068609604      1.4621826971735778    203526   
211   201143    0.37843044876673804       1.277668801298337    201143   
446   201599     0.9643061686157518      0.6096249505772804    201599   
126   203953     0.7428822870634291       0.815305009121352    203953   
285     2733     0.7386394447442929      0.8109439798035029      2733   
92   1626273    -0.0787271031823476      1.5707434341634514   1626273   
376   203503   -0.12003676718303119      1.5919430074038063    203503   
300   101133   -0.24614161117316355      1.7010975964285544    101133   
237   201936     1.2072630179962267      0.2475604264357446    201936   
307   200768     0.7426361337129879      0.7120132443963622    200768   
60    202066    -0.5421114888233356      1.9817938334084404    202066   
354   203481     0.5499247972784587      0.8882357016069932    203481   
390   203141     2.2079017451594525     -0.7870308351879381    203141   
66    202700    0.44047028929144083      0.9744824385828311    202700   
436   203991   -0.04954955574915776      1.4276501412849758    20399

#  Below here is code from first version

1 (make), 2(miss), 3 (ft), 4 (reb), 5 (tov), 6(foul), 7(violation), 8(sub), 9(timeout), 10 (jumpball), 11 (ejection)

In [166]:
lineups = []
home_ball_lineup = []
away_ball_lineup = []

#If the team with the lower TEAM_ID number has the ball
for i in stints[(stints['HOME_TEAM'] < stints['AWAY_TEAM']) & (stints['HA'] == 'H') & (stints['STINT_POSS'] >= 1)]['stint']:
    home_offense = OrderedDict({f"{j},offense": 1 for j in i.split(', ')[:5]})
    away_defense = OrderedDict({f"{j},defense": (-1) for j in i.split(', ')[5:]})
    home_ball_lineup = home_offense.copy()
    home_ball_lineup.update(away_defense)
    lineups.append(home_ball_lineup)

for i in stints[(stints['AWAY_TEAM'] < stints['HOME_TEAM']) & (stints['HA'] == 'A') & (stints['STINT_POSS'] >= 1)]['stint']:
    away_offense = OrderedDict({f"{j},offense": 1 for j in i.split(', ')[:5]})
    home_defense = OrderedDict({f"{j},defense": (-1) for j in i.split(', ')[5:]})
    away_ball_lineup = away_offense.copy()
    away_ball_lineup.update(home_defense)
    lineups.append(away_ball_lineup)

#If the team with the higher TEAM_ID number has the ball
for i in stints[(stints['HOME_TEAM'] > stints['AWAY_TEAM']) & (stints['HA'] == 'H') & (stints['STINT_POSS'] >= 1)]['stint']:
    home_offense_b = OrderedDict({f"{k},offense": 1 for k in [s for s in i.split(', ')[::-1]][:5]})
    away_defense_b = OrderedDict({f"{k},defense": (-1) for k in [s for s in i.split(', ')[::-1]][5:]})
    home_ball_lineup_b = home_offense_b.copy()
    home_ball_lineup_b.update(away_defense_b)
    lineups.append(home_ball_lineup_b)

for i in stints[(stints['AWAY_TEAM'] > stints['HOME_TEAM']) & (stints['HA'] == 'A') & (stints['STINT_POSS'] >= 1)]['stint']:
    away_offense_b = OrderedDict({f"{k},offense": 1 for k in [s for s in i.split(', ')[::-1]][:5]})
    home_defense_b = OrderedDict({f"{k},defense": (-1) for k in [s for s in i.split(', ')[::-1]][5:]})
    away_ball_lineup_b = away_offense_b.copy()
    away_ball_lineup_b.update(home_defense_b)
    lineups.append(away_ball_lineup)

orating = []
for s in stints[stints['STINT_POSS'] >= 1]['ORATING']:
    orating.append(s)

possessions = []
for s in stints[stints['STINT_POSS'] >= 1]['STINT_POSS']:
    possessions.append(s)



#print(lineups[0:10])
#print(len(lineups))
#print(len(stints['stint']))
#print(len(orating))
#print(len(possessions))
#print(orating[0:25])
#print(possessions[0:25])

200794,offense  201143,offense  201952,offense  203145,offense  \
0             1.0             1.0             1.0             1.0   
1             NaN             1.0             NaN             NaN   
2             NaN             1.0             NaN             NaN   
3             NaN             1.0             NaN             1.0   
4             NaN             1.0             1.0             NaN   

   2594,offense  101141,defense  202694,defense  202704,defense  \
0           1.0            -1.0            -1.0            -1.0   
1           NaN             NaN            -1.0             NaN   
2           NaN             NaN            -1.0             NaN   
3           1.0            -1.0             NaN             NaN   
4           1.0            -1.0             NaN             NaN   

   203083,defense  203484,defense  200757,offense  201168,offense  \
0            -1.0            -1.0             NaN             NaN   
1             NaN            -1.0             1.0             1.0   
2             NaN             NaN             1.0             NaN   
3             NaN             NaN             NaN             NaN   
4            -1.0             NaN             1.0             NaN   

   203471,offense  203934,offense  1626169,defense  203382,defense  \
0             NaN             NaN              NaN             NaN   
1             1.0             1.0             -1.0            -1.0   
2             1.0             1.0             -1.0            -1.0   
3             1.0             NaN             -1.0            -1.0   
4             NaN             NaN             -1.0             NaN   

   2581,defense  203488,offense  201975,defense  203118,offense  \
0           NaN             NaN             NaN             NaN   
1          -1.0             NaN             NaN             NaN   
2          -1.0             1.0            -1.0             NaN   
3          -1.0             1.0            -1.0             NaN   
4          -1.0             1.0            -1.0             NaN   

   202696,offense  202699,offense  203095,offense  203506,offense  \
0             NaN             NaN             NaN             NaN   
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   203901,offense  101162,defense  202322,defense  203078,defense  \
0             NaN             NaN             NaN             NaN   
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   203490,defense  2743,defense  1626209,offense  201228,offense  \
0             NaN           NaN              NaN             NaN   
1             NaN           NaN              NaN             NaN   
2             NaN           NaN              NaN             NaN   
3             NaN           NaN              NaN             NaN   
4             NaN           NaN              NaN             NaN   

   203932,offense  201196,defense  202390,defense  2400,defense  2403,defense  \
0             NaN             NaN             NaN           NaN           NaN   
1             NaN             NaN             NaN           NaN           NaN   
2             NaN             NaN             NaN           NaN           NaN   
3             NaN             NaN             NaN           NaN           NaN   
4             NaN             NaN             NaN           NaN           NaN   

   203473,offense  201160,offense  101135,offense  202340,offense  \
0             NaN             NaN             NaN             NaN   
1             NaN             NaN             NaN             NaN   
2            

In [84]:
y = csr_matrix(orating)
y = y.transpose()
print(X.shape[0])
print(y.shape[0])

np.set_printoptions(threshold=np.inf)
v = DictVectorizer(sparse = True)    
X = v.fit_transform(lineups)

#v.inverse_transform(X[0:25])
x = X.toarray()
print((x == 1).sum())
print((x == (-1)).sum())

60227
60227
301019
301266


In [85]:
v = DictVectorizer()    
X = v.fit_transform(lineups)


lambdas = [.01, .025, .05, .1]
samples = X.shape[0]
alphas = [(l * samples / 2) for l in lambdas]

#clf = linear_model.RidgeCV(alphas = (np.array([0.01, 0.1, 1.0, 10, 100, 500, 750])), cv = 5)
clf = linear_model.RidgeCV(alphas = alphas, cv = 5)
model = clf.fit(X, orating, sample_weight = possessions)
#model = clf.fit(X, stints['ORATING'], sample_weight = stints['STINT_POSS'])

In [89]:
unfiltered_RAPM.sort_values(["RAPM"])
#RAPM.sort_values('O-RAPM', ascending = False)

id       ID     side      RAPM
679   203498,offense   203498  offense -4.543681
93   1626167,offense  1626167  offense -4.150355
177     1938,offense     1938  offense -3.973479
956      977,offense      977  offense -3.764554
883     2544,offense     2544  offense -3.612511
928     2738,offense     2738  offense -3.603481
885     2546,offense     2546  offense -3.565708
950     2772,offense     2772  offense -3.536898
653   203477,offense   203477  offense -3.473163
906     2581,offense     2581  offense -3.450518
9     101108,offense   101108  offense -3.437461
924     2734,offense     2734  offense -3.421663
525   202738,offense   202738  offense -3.402494
213   201143,offense   201143  offense -3.360193
201   200782,offense   200782  offense -3.328230
389   202083,offense   202083  offense -3.294462
249   201188,offense   201188  offense -3.293164
263   201565,offense   201565  offense -3.253072
659   203484,offense   203484  offense -3.229653
505   202710,offense   202710  offense -3.167257
5     101106,offense   101106  offense -3.150540
565   203095,offense   203095  offense -3.106930
940     2754,offense     2754  offense -3.103886
369   201967,offense   201967  offense -3.073605
189   200755,offense   200755  offense -3.068396
904     2571,offense     2571  offense -3.052009
299   201585,offense   201585  offense -3.035679
267   201567,offense   201567  offense -3.019229
281   201575,offense   201575  offense -3.019084
445   202389,offense   202389  offense -2.976572
335   201939,offense   201939  offense -2.969093
435   202355,offense   202355  offense -2.956906
599   203138,offense   203138  offense -2.927127
733   203897,offense   203897  offense -2.921110
775   203932,offense   203932  offense -2.904682
571   203099,offense   203099  offense -2.901848
693   203507,offense   203507  offense -2.894241
313   201599,offense   201599  offense -2.885531
269   201568,offense   201568  offense -2.877997
781   203935,offense   203935  offense -2.865656
735   203898,offense   203898  offense -2.864213
741   203903,offense   203903  offense -2.862325
27    101135,offense   101135  offense -2.861504
329   201935,offense   201935  offense -2.842189
203   200794,offense   200794  offense -2.841043
353   201952,offense   201952  offense -2.833460
723     2037,offense     2037  offense -2.812304
235   201160,offense   201160  offense -2.785887
691   203506,offense   203506  offense -2.785584
441   202379,offense   202379  offense -2.763238
489   202699,offense   202699  offense -2.750996
363   201961,offense   201961  offense -2.749198
912     2590,offense     2590  offense -2.698326
77   1626157,offense  1626157  offense -2.687989
627   203457,offense   203457  offense -2.682575
359   201959,offense   201959  offense -2.659650
11    101109,offense   101109  offense -2.643744
295   201583,offense   201583  offense -2.641816
451   202397,offense   202397  offense -2.636360
193   200765,offense   200765  offense -2.599857
337   201941,offense   201941  offense -2.588608
561   203093,offense   203093  offense -2.583973
417   202337,offense   202337  offense -2.575802
7     101107,offense   101107  offense -2.561386
861     2225,offense     2225  offense -2.561065
481   202694,offense   202694  offense -2.560673
81   1626159,offense  1626159  offense -2.528892
711   203530,offense   203530  offense -2.526754
471   202689,offense   202689  offense -2.488068
569   203098,offense   203098  offense -2.465580
37    101150,offense   101150  offense -2.459601
703   203521,offense   203521  offense -2.443969
391   202087,offense   202087  offense -2.382326
187   200752,offense   200752  offense -2.374300
914     2592,offense     2592  offense -2.370561
595   203124,offense   203124  offense -2.365832
126  1626196,offense  1626196  offense -2.343156
217   201145,offense   201145  offense -2.319394
405   202329,offense   202329  offense -2.299617
87   1626163,offense  1626163  offense -2.299548
645   203471,offense   203471  offense -2.273782


In [50]:
RAPM.sort_values('O-RAPM', ascending = False)

ID               player_name    O-RAPM    D-RAPM      RAPM   SEASON
9     101127              Jarrett Jack  1.718673  0.096420  1.815092  2014-15
19    101179              Ronnie Price  1.645438  0.498552  2.143990  2014-15
207   202338            Kevin Seraphin  1.591930  1.250510  2.842440  2014-15
254   202714              Shelvin Mack  1.590080  1.547084  3.137164  2014-15
437     2446             Rasual Butler  1.242914  1.148376  2.391290  2014-15
169   201945          Gerald Henderson  1.073432  2.044209  3.117641  2014-15
55   1626179              Terry Rozier  0.855376  0.073345  0.928721  2014-15
444     2550              Kirk Hinrich  0.815445 -0.724075  0.091370  2014-15
373   203915         Spencer Dinwiddie  0.738908  0.837917  1.576825  2014-15
332   203490           Otto Porter Jr.  0.686626  1.895853  2.582479  2014-15
448     2564                Boris Diaw  0.630718  2.980934  3.611651  2014-15
261   202962                Greg Smith  0.620196 -0.652696 -0.032500  2014-15
70   1626246          Boban Marjanovic  0.614197  0.822365  1.436562  2014-15
290   203112                Quincy Acy  0.566568  0.989337  1.555905  2014-15
43   1626166             Cameron Payne  0.562506  0.850347  1.412853  2014-15
277   203092              Tyler Zeller  0.455512 -0.119536  0.335976  2014-15
79      1713              Vince Carter  0.414069  1.536155  1.950224  2014-15
59   1626191          Chris McCullough  0.411946 -0.735963 -0.324017  2014-15
429     2365            Chris Andersen  0.397604  0.108839  0.506442  2014-15
302   203148             Brian Roberts  0.394005  0.351271  0.745276  2014-15
383   203928            Cory Jefferson  0.356259  0.156595  0.512854  2014-15
120   201171               Carl Landry  0.349015  2.902527  3.251541  2014-15
303   203156                Henry Sims  0.345071  0.880804  1.225875  2014-15
78   1627362             Briante Weber  0.340995  0.334262  0.675257  2014-15
164   201937               Ricky Rubio  0.338591  1.320987  1.659578  2014-15
52   1626175             Jordan Mickey  0.299150  0.217019  0.516169  2014-15
452     2585             Zaza Pachulia  0.271977  2.474446  2.746423  2014-15
83      1882               Elton Brand  0.247267  1.157364  1.404631  2014-15
361   203893                Russ Smith  0.247039  0.395817  0.642856  2014-15
284   203100               Tony Wroten  0.242357  0.787174  1.029531  2014-15
97    200779               Steve Novak  0.233731 -0.532877 -0.299147  2014-15
420   204067              Jarell Eddie  0.228879  0.354260  0.583138  2014-15
8     101126             Nate Robinson  0.226529  0.136340  0.362869  2014-15
22    101236               Chuck Hayes  0.216298 -0.182236  0.034062  2014-15
144   201582             Alexis Ajinca  0.206427  1.955892  2.162319  2014-15
369   203906              Devyn Marble  0.198877  0.968959  1.167836  2014-15
117   201166              Aaron Brooks  0.176212  2.090816  2.267028  2014-15
397   203950            Jarnell Stokes  0.154039  0.013737  0.167776  2014-15
357   203613          Jonathon Simmons  0.152774  0.605328  0.758101  2014-15
325   203481               Jeff Withey  0.152362  1.175824  1.328187  2014-15
114   201158           Marco Belinelli  0.148601  3.838350  3.986951  2014-15
103   201142              Kevin Durant  0.140818  2.349621  2.490438  2014-15
82      1737             Nazr Mohammed  0.136880  0.133022  0.269903  2014-15
402   203956              Mitch McGary  0.123515  0.126220  0.249735  2014-15
153   201596            Mario Chalmers  0.107147  2.836372  2.943519  2014-15
57   1626183            Branden Dawson  0.095577  0.440806  0.536383  2014-15
358   203648    Thanasis Antetokounmpo  0.081013  0.004764  0.085777  2014-15
274   203088          Kendall Marshall  0.076658  1.070095  1.146753  2014-15
194   202322                 John Wall  0.072993  0.676501  0.749494  2014-15
315   203461           Anthony Bennett  0.050170  0.106049  0.156219  2014-15
152   201593            Nikola P

In [ ]:
RAPM.sort_values('RAPM', ascending = False)

In [ ]:
RAPM.sort_values('RAPM', ascending = False)

In [ ]:
RAPM[RAPM['RAPM'].isna()]

In [41]:
#Formatting ['ordered_stint']

#[k for k in [s for s in [i.split(', ') for i in stints['stint']]][::-1]]
#[i.split(', ') for i in stints['stint']]
#print(len([i.split(', ') for i in stints['stint']]))


#[s[::-1] for s in [i.split(', ') for i in stints['stint']]]
#[s[2::-1] for s in [i.split(', ') for i in stints['stint']]]

#type([s*3 for s in [i.split(', ') for i in stints['stint']]])
#print(len([s for s in [i.split(', ') for i in stints['stint']]]))
#[k*2 for k in [s for s in [i.split(', ') for i in stints['stint']]]]
#print(len([k for k in [s for s in [i.split(', ') for i in stints['stint']]]]))

In [ ]:
#stints.head()
#print(stints[stints['HOME_TEAM'] > stints['AWAY_TEAM']].count())
#print(stints[stints['AWAY_TEAM'] > stints['HOME_TEAM']].count())
#print(stints[stints['AWAY_TEAM'] == stints['HOME_TEAM']].count())
#len(stints[stints['stint'].str.contains('1628378')])
#len(stints[(stints['STINT_POSS'] == 0) & (stints['STINT_PTS'] > 0)])
#len(stints[stints['STINT_POSS'] == 0])
#stints[stints['STINT_POSS'] == 0].head(20)
#len(stints[(stints['STINT_POSS'] == 0) & (stints['STINT_PTS'] > 0)]['stint'].drop_duplicates())

#pbp_df[(pbp_df['stint'] == '201569, 201600, 201936, 201950, 203076, 202696, 202699, 203095, 203901, 203932')
#      & (pbp_df['game_id'] == '21400001') & (pbp_df['PERIOD'] == '2')]

#pbp[(pbp['GAME_ID'] == '21400001') & (pbp['PERIOD'] == '2')].sort_values('EVENTNUM')


#pbp_df[(pbp_df['stint'] == '101141, 101141, 201162, 201162, 202336, 202336, 202688, 202688, 203953, 203953, 101139, 101139, 201579, 201579, 202388, 202388, 203524, 203524, 2449, 2449')].sort_values('e')

#print(pbp_df[(pbp_df['homedescription'] != '') & (pbp_df['visitordescription'] != '')]['EVENTMSGTYPE'].value_counts())
#pbp_df[(pbp_df['homedescription'] != '') & (pbp_df['visitordescription'] != '') & (pbp_df['EVENTMSGTYPE'] == '2')].head()

#pbp_df[(pbp_df['EVENTMSGTYPE'] == '3') 
#                          & (pbp_df['DESCRIPTION'].str.contains('MISS'))
#                          & (pbp_df['NEXT_VISITORDESCRIPTION'].str.contains('REBOUND'))
#                          & (pbp_df['VISITORDESCRIPTION'] == '')
#                          & (pbp_df['POSS'] == 0)].head()

lineups = []
home_ball_lineup = []
away_ball_lineup = []

#If the team with the lower TEAM_ID number has the ball
for i in stints[(stints['HOME_TEAM'] < stints['AWAY_TEAM']) & (stints['HA'] == 'H') & (stints['STINT_POSS'] >= 1)]['stint']:
    home_offense = [j for j in i.split(', ')][:5]
    away_defense = [j for j in i.split(', ')][5:]
    home_ball_lineup = home_offense.copy()
    home_ball_lineup.extend(away_defense)
    lineups.append(home_ball_lineup)

for i in stints[(stints['AWAY_TEAM'] < stints['HOME_TEAM']) & (stints['HA'] == 'A') & (stints['STINT_POSS'] >= 1)]['stint']:
    away_offense = [j for j in i.split(', ')][:5]
    home_defense = [j for j in i.split(', ')][5:]
    away_ball_lineup = away_offense.copy()
    away_ball_lineup.extend(home_defense)
    lineups.append(away_ball_lineup)

#If the team with the higher TEAM_ID number has the ball
for i in stints[(stints['HOME_TEAM'] > stints['AWAY_TEAM']) & (stints['HA'] == 'H') & (stints['STINT_POSS'] >= 1)]['stint']:
    home_offense_b = [k for k in [s for s in i.split(', ')[::-1]]][:5]
    away_defense_b = [k for k in [s for s in i.split(', ')[::-1]]][5:]
    home_ball_lineup_b = home_offense_b.copy()
    home_ball_lineup_b.extend(away_defense_b)
    lineups.append(home_ball_lineup_b)

for i in stints[(stints['AWAY_TEAM'] > stints['HOME_TEAM']) & (stints['HA'] == 'A') & (stints['STINT_POSS'] >= 1)]['stint']:
    away_offense_b = [k for k in [s for s in i.split(', ')[::-1]]][:5]
    home_defense_b = [k for k in [s for s in i.split(', ')[::-1]]][5:]
    away_ball_lineup_b = away_offense_b.copy()
    away_ball_lineup_b.extend(home_defense_b)
    lineups.append(away_ball_lineup)

lineup_df = pd.DataFrame.from_records(lineups)
lineup_df = lineup_df.rename(columns = {0:'offense_1', 1:'offense_2', 2:'offense_3', 3:'offense_4', 4:'offense_5',
                                       5:'defense_1', 6:'defense_2', 7:'defense_3', 8:'defense_4', 9:'defense_5'})
lineup_df.head()


#pbp_df[pbp_df['stint'] == '101112, 200768, 201942, 2449'].sort_values(['e'])
l = []
for i in pbp_df['stint']:
    if len([j for j in i.split(', ')]) < 10:
        l.append(i)
len(l)

In [ ]:
#array_of_players = np.array(set_of_players)
#array_of_players = array_of_players.reshape(1,len(array_of_players))
#type(array_of_players[0][0])
#set_of_players.index('1626154')

arr = lineups_only.to_numpy()
arr = arr.T
type(lineups_only)

In [806]:
unfiltered_RAPM = []

offensive = np.transpose(model.coef_[:, 0:num_of_players])
defensive = np.transpose(model.coef_[:, num_of_players:])

for pid in player_ids:
    unfiltered_RAPM.append((pid, model.coef_[player_ids.index(pid)]))

unfiltered_RAPM = pd.DataFrame.from_records(unfiltered_RAPM, columns = ['id','RAPM'])
unfiltered_RAPM.insert(loc = 1, column = 'ID', value = ([i[0] for i in unfiltered_RAPM['id'].str.split(',')]))
unfiltered_RAPM.insert(loc = 2, column = 'side', value = ([i[1] for i in unfiltered_RAPM['id'].str.split(',')]))

RAPM = unfiltered_RAPM.merge(right = roster, how = 'left', left_on = 'ID', right_on = 'person_id')

#O-RAPM / D-RAPM
O_RAPM = RAPM[RAPM['side'] == 'offense'][['ID','RAPM']]
O_RAPM = O_RAPM.rename(columns = {"RAPM":"O-RAPM"})
D_RAPM = RAPM[RAPM['side'] == 'defense'][['ID','RAPM']]
D_RAPM = D_RAPM.rename(columns = {"RAPM":"D-RAPM"})

components = O_RAPM.merge(right = D_RAPM, how = 'left', on = ['ID'])
names = RAPM[['ID','player_name']].drop_duplicates()
RAPM = names.merge(right = components, how = 'left', on = ['ID'])
RAPM['RAPM'] = RAPM['O-RAPM'] + RAPM['D-RAPM']
RAPM['SEASON'] = '2014-15'
RAPM.to_csv("C://Users/gsteele/Other/RAPM/2014_15_RAPM.csv")

NameError: name 'v' is not defined

In [47]:
print(len(unfiltered_RAPM[unfiltered_RAPM['side'] == 'defense']))
print(len(unfiltered_RAPM)/2)

475
474.5
